File: sdk/go/client_test.go
package peakauth

import (
	"crypto/rand"
	"crypto/rsa"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"math/big"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

func TestGeneratePKCE(t *testing.T) {
	pkce, err := GeneratePKCE(64)
	if err != nil {
		t.Fatalf("GeneratePKCE falló: %v", err)
	}

	if pkce.CodeVerifier == "" {
		t.Fatal("CodeVerifier no debe estar vacío")
	}
	if pkce.CodeChallenge == "" {
		t.Fatal("CodeChallenge no debe estar vacío")
	}
	if len(pkce.CodeVerifier) < 43 {
		t.Fatalf("CodeVerifier demasiado corto: %d", len(pkce.CodeVerifier))
	}
}

func TestGetAuthorizationURL(t *testing.T) {
	client, err := New(Config{
		IssuerURL:   "https://auth.example.com",
		ClientID:    "test-app",
		RedirectURI: "https://my-app.com/callback",
	})
	if err != nil {
		t.Fatalf("New falló: %v", err)
	}

	u, err := client.GetAuthorizationURL("state-123", "challenge-456")
	if err != nil {
		t.Fatalf("GetAuthorizationURL falló: %v", err)
	}

	expected := "https://auth.example.com/oauth/authorize?client_id=test-app&code_challenge=challenge-456&code_challenge_method=S256&redirect_uri=https%3A%2F%2Fmy-app.com%2Fcallback&response_type=code&state=state-123"
	if u != expected {
		t.Fatalf("URL generada inesperada.\nEsperada: %s\nObtenida: %s", expected, u)
	}
}

func TestVerifyTokenWithJWKS(t *testing.T) {
	// 1. Generar par de claves RSA para el test
	privKey, err := rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		t.Fatalf("error generando clave RSA: %v", err)
	}

	nStr := base64.RawURLEncoding.EncodeToString(privKey.PublicKey.N.Bytes())
	eStr := base64.RawURLEncoding.EncodeToString(big.NewInt(int64(privKey.PublicKey.E)).Bytes())
	kid := "test-key-id"

	// 2. Levantar servidor mock de Peak Auth que expone JWKS
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path == "/.well-known/jwks.json" {
			w.Header().Set("Content-Type", "application/json")
			_ = json.NewEncoder(w).Encode(jwksResponse{
				Keys: []jwk{
					{
						Kty: "RSA",
						Alg: "RS256",
						Use: "sig",
						Kid: kid,
						N:   nStr,
						E:   eStr,
					},
				},
			})
			return
		}
		http.NotFound(w, r)
	}))
	defer server.Close()

	client, err := New(Config{
		IssuerURL:      server.URL,
		ClientID:       "my-app",
		ExpectedIssuer: "peak-auth",
	})
	if err != nil {
		t.Fatalf("New falló: %v", err)
	}

	// 3. Crear token válido
	claims := Claims{
		Username:    "bruno@example.com",
		AppID:       "my-app",
		Roles:       []string{"ADMIN", "USER"},
		MfaVerified: true,
		TokenType:   "access",
		RegisteredClaims: jwt.RegisteredClaims{
			Subject:   "101",
			Issuer:    "peak-auth",
			Audience:  jwt.ClaimStrings{"my-app"},
			ExpiresAt: jwt.NewNumericDate(time.Now().Add(time.Hour)),
			IssuedAt:  jwt.NewNumericDate(time.Now()),
		},
	}

	token := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
	token.Header["kid"] = kid
	tokenStr, err := token.SignedString(privKey)
	if err != nil {
		t.Fatalf("error firmando token: %v", err)
	}

	// 4. Validar token
	verifiedClaims, err := client.VerifyToken(tokenStr)
	if err != nil {
		t.Fatalf("VerifyToken falló: %v", err)
	}

	if verifiedClaims.Subject != "101" || verifiedClaims.Username != "bruno@example.com" {
		t.Fatalf("claims incorrectos: %+v", verifiedClaims)
	}

	// 5. Validar que falle con audiencia equivocada
	wrongAudClaims := claims
	wrongAudClaims.Audience = jwt.ClaimStrings{"otra-app"}
	wrongToken := jwt.NewWithClaims(jwt.SigningMethodRS256, wrongAudClaims)
	wrongToken.Header["kid"] = kid
	wrongTokenStr, _ := wrongToken.SignedString(privKey)

	if _, err := client.VerifyToken(wrongTokenStr); err == nil {
		t.Fatal("VerifyToken debía fallar por mismatch de audiencia")
	}
}

func TestGinAndHTTPMiddleware(t *testing.T) {
	privKey, _ := rsa.GenerateKey(rand.Reader, 2048)
	nStr := base64.RawURLEncoding.EncodeToString(privKey.PublicKey.N.Bytes())
	eStr := base64.RawURLEncoding.EncodeToString(big.NewInt(int64(privKey.PublicKey.E)).Bytes())

	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(jwksResponse{
			Keys: []jwk{
				{Kty: "RSA", Alg: "RS256", Use: "sig", Kid: "k1", N: nStr, E: eStr},
			},
		})
	}))
	defer server.Close()

	client, _ := New(Config{
		IssuerURL:      server.URL,
		ClientID:       "test-app",
		ExpectedIssuer: "peak-auth",
	})

	// Token con rol USER (no ADMIN)
	token := jwt.NewWithClaims(jwt.SigningMethodRS256, Claims{
		Username: "user1",
		AppID:    "test-app",
		Roles:    []string{"USER"},
		RegisteredClaims: jwt.RegisteredClaims{
			Subject:   "1",
			Issuer:    "peak-auth",
			Audience:  jwt.ClaimStrings{"test-app"},
			ExpiresAt: jwt.NewNumericDate(time.Now().Add(time.Hour)),
		},
	})
	token.Header["kid"] = "k1"
	tokenStr, _ := token.SignedString(privKey)

	// Test HTTP Middleware (Standard Library net/http)
	stdHandler := client.HTTPMiddleware("USER")(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		claims, ok := ClaimsFromContext(r.Context())
		if !ok || claims.Username != "user1" {
			http.Error(w, "error en claims", http.StatusInternalServerError)
			return
		}
		w.WriteHeader(http.StatusOK)
	}))

	// 1. Petición con rol adecuado (USER) -> 200 OK
	w1 := httptest.NewRecorder()
	req1, _ := http.NewRequest(http.MethodGet, "/", nil)
	req1.Header.Set("Authorization", fmt.Sprintf("Bearer %s", tokenStr))
	stdHandler.ServeHTTP(w1, req1)

	if w1.Code != http.StatusOK {
		t.Fatalf("se esperaba 200 OK en HTTPMiddleware, obtenido: %d", w1.Code)
	}

	// 2. Petición con rol faltante (ADMIN) -> 403 Forbidden
	adminHandler := client.HTTPMiddleware("ADMIN")(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusOK)
	}))
	w2 := httptest.NewRecorder()
	req2, _ := http.NewRequest(http.MethodGet, "/", nil)
	req2.Header.Set("Authorization", fmt.Sprintf("Bearer %s", tokenStr))
	adminHandler.ServeHTTP(w2, req2)

	if w2.Code != http.StatusForbidden {
		t.Fatalf("se esperaba 403 Forbidden en HTTPMiddleware para rol ADMIN, obtenido: %d", w2.Code)
	}
}

File: sdk/typescript/src/client.test.ts
import { describe, it, before, after } from 'node:test';
import assert from 'node:assert';
import http from 'node:http';
import * as jose from 'jose';
import { generatePKCE } from './pkce.js';
import { PeakAuthClient } from './client.js';

describe('Peak Auth TypeScript SDK', () => {
  describe('PKCE Generator', () => {
    it('debe generar un par PKCE válido con code_verifier y code_challenge', async () => {
      const pkce = await generatePKCE();
      assert.ok(pkce.codeVerifier);
      assert.ok(pkce.codeChallenge);
      assert.ok(pkce.codeVerifier.length >= 43 && pkce.codeVerifier.length <= 128);

      // Verificamos que code_challenge sea el hash SHA-256 en base64url
      const encoder = new TextEncoder();
      const digest = await crypto.subtle.digest('SHA-256', encoder.encode(pkce.codeVerifier));
      const expectedChallenge = Buffer.from(digest).toString('base64url');
      assert.strictEqual(pkce.codeChallenge, expectedChallenge);
    });

    it('debe generar verifiers aleatorios únicos en cada invocación', async () => {
      const p1 = await generatePKCE();
      const p2 = await generatePKCE();
      assert.notStrictEqual(p1.codeVerifier, p2.codeVerifier);
    });
  });

  describe('PeakAuthClient', () => {
    it('debe requerir issuerUrl y clientId', () => {
      assert.throws(() => new PeakAuthClient({ issuerUrl: '', clientId: 'app' }), /issuerUrl es requerido/);
      assert.throws(() => new PeakAuthClient({ issuerUrl: 'http://localhost', clientId: '' }), /clientId es requerido/);
    });

    it('debe construir la URL de autorización correctamente con parámetros PKCE', () => {
      const client = new PeakAuthClient({
        issuerUrl: 'https://auth.example.com',
        clientId: 'my-app',
        redirectUri: 'https://app.com/callback',
      });

      const url = client.getAuthorizationUrl({
        state: 'xyz123',
        codeChallenge: 'challenge456',
        scope: 'openid profile',
      });

      const parsed = new URL(url);
      assert.strictEqual(parsed.origin, 'https://auth.example.com');
      assert.strictEqual(parsed.pathname, '/oauth/authorize');
      assert.strictEqual(parsed.searchParams.get('client_id'), 'my-app');
      assert.strictEqual(parsed.searchParams.get('redirect_uri'), 'https://app.com/callback');
      assert.strictEqual(parsed.searchParams.get('response_type'), 'code');
      assert.strictEqual(parsed.searchParams.get('state'), 'xyz123');
      assert.strictEqual(parsed.searchParams.get('code_challenge'), 'challenge456');
      assert.strictEqual(parsed.searchParams.get('code_challenge_method'), 'S256');
      assert.strictEqual(parsed.searchParams.get('scope'), 'openid profile');
    });
  });

  describe('Token Verification contra JWKS Mock', () => {
    let server: http.Server;
    let serverUrl: string;
    let keyPair: jose.GenerateKeyPairResult;
    const kid = 'test-rsa-kid-1';

    before(async () => {
      // 1. Generar par de claves RSA RS256
      keyPair = await jose.generateKeyPair('RS256');
      const publicJwk = await jose.exportJWK(keyPair.publicKey);
      publicJwk.kid = kid;
      publicJwk.use = 'sig';
      publicJwk.alg = 'RS256';

      // 2. Levantar servidor HTTP mock de JWKS
      await new Promise<void>((resolve) => {
        server = http.createServer((req, res) => {
          if (req.url === '/.well-known/jwks.json') {
            res.writeHead(200, { 'Content-Type': 'application/json' });
            res.end(JSON.stringify({ keys: [publicJwk] }));
            return;
          }
          res.writeHead(404);
          res.end();
        });
        server.listen(0, '127.0.0.1', () => {
          const addr = server.address() as { port: number };
          serverUrl = `http://127.0.0.1:${addr.port}`;
          resolve();
        });
      });
    });

    after(async () => {
      await new Promise<void>((resolve) => server.close(() => resolve()));
    });

    it('debe validar exitosamente un token firmado con la clave del JWKS', async () => {
      const client = new PeakAuthClient({
        issuerUrl: serverUrl,
        clientId: 'target-app',
        expectedIssuer: 'peak-auth',
      });

      // Firmar token JWT con la clave privada y cabecera kid
      const jwt = await new jose.SignJWT({
        username: 'bruno@peak.test',
        app_id: 'target-app',
        roles: ['ADMIN', 'USER'],
      })
        .setProtectedHeader({ alg: 'RS256', kid })
        .setIssuer('peak-auth')
        .setAudience('target-app')
        .setSubject('101')
        .setIssuedAt()
        .setExpirationTime('1h')
        .sign(keyPair.privateKey);

      const claims = await client.verifyToken(jwt);
      assert.strictEqual(claims.sub, '101');
      assert.strictEqual(claims.username, 'bruno@peak.test');
      assert.strictEqual(claims.app_id, 'target-app');
      assert.deepStrictEqual(claims.roles, ['ADMIN', 'USER']);
    });

    it('debe rechazar un token con audiencia incorrecta', async () => {
      const client = new PeakAuthClient({
        issuerUrl: serverUrl,
        clientId: 'different-app',
        expectedIssuer: 'peak-auth',
      });

      const jwt = await new jose.SignJWT({
        username: 'hacker@test.com',
        app_id: 'target-app',
        roles: ['USER'],
      })
        .setProtectedHeader({ alg: 'RS256', kid })
        .setIssuer('peak-auth')
        .setAudience('target-app')
        .setExpirationTime('1h')
        .sign(keyPair.privateKey);

      await assert.rejects(async () => {
        await client.verifyToken(jwt);
      }, /unexpected "aud" claim value/);
    });
  });
});
